sympy/solvers/polysys.py

```diff
@@ -429,3 +439,389 @@ def solve_triangulated(polys, *gens, **args):
+          being a Poly instance and conditions being a tuple of Poly instances
+        - constant_conds is a list of Poly instances representing conditions that
+          must be satisfied for any solution to exist
+
+    Examples
+    ========
+
+    >>> from sympy import symbols, Poly, ZZ
+    >>> from sympy.solvers.polysys import factor_system_poly
+    >>> a, b, c, x = symbols('a b c x')
+    >>> p1 = Poly((a - 1)*(x - 2), x, domain=ZZ[a,b,c])
+    >>> p2 = Poly((b - 3)*(x - 2), x, domain=ZZ[a,b,c])
+    >>> p3 = Poly(c, x, domain=ZZ[a,b,c])
+
+    The equation to be solved for x is ``x - 2 = 0`` provided either
+    of the two conditions on the parameters ``a`` and ``b`` is nonzero
+    and the constant parameter ``c`` should be zero.
+
+    >>> systems, constant_conds = factor_system_poly([p1, p2, p3])
+    >>> systems
+    [[(Poly(x - 2, x, domain='ZZ[a,b,c]'), (Poly(a - 1, x, domain='ZZ[a,b,c]'), Poly(b - 3, x, domain='ZZ[a,b,c]')))]]
+
+    >>> constant_conds
+    [Poly(c, x, domain='ZZ[a,b,c]')]
+
+    Notes
+    =====
+
+    - This is the core routine used by higher-level functions factor_system_cond,
+      factor_poly_system, and factor_system_bool
+    - Returns empty systems list ([]) when no solution exists
+    - Returns [[]] (list containing empty list) when any value is a solution
+    """
+    if not all(isinstance(poly, Poly) for poly in polys):
+        raise TypeError("polys should be a list of Poly instances")
+    if not polys:
+        return [[]], []
+    else:
+        base_domain = polys[0].domain
+        base_gens = polys[0].gens
+        if not all(poly.domain == base_domain and poly.gens == base_gens for poly in polys[1:]):
+            raise DomainError("All polynomials must have the same domain and generators")
+
+    constant_eqs = []
+    eqs_factors = []
+    conds_factor = defaultdict(list)
+
+    for poly in polys:
+        constant, factors_mult = poly.factor_list()
+
+        factors = [f for f, m in factors_mult]
+
+        if factors:
+            eqs_factors.append(factors)
+            if constant.is_zero is not False:
+                constp = constant.as_poly(base_gens, domain=base_domain)
+                for f in factors:
+                    if constp not in conds_factor[f]:
+                        conds_factor[f].append(constp)
+        elif constant.is_zero is True:
+            pass
+        elif constant.is_zero is False:
+            return ([], [])
+        else:
+            constp = constant.as_poly(base_gens, domain=base_domain)
+            constant_eqs.append(constp)
+
+    fac2conds = {f: tuple(conds) for f, conds in conds_factor.items()}
+
+    cnf = [[(f, fac2conds.get(f, ())) for f in eq] for eq in eqs_factors]
+
+    dnf = _cnf2dnf(cnf)
+
+    return dnf, constant_eqs
+
+
+def _has_common_variables(factor1, factor2):
+    """Helper function to check if two polynomial factors share any variables"""
+    return bool(factor1.free_symbols.intersection(factor2.free_symbols))
```

Comment:
I don't think that free symbols should be used here at all. This should be checking the generators of the polynomials or something.

Reply:
I had a similar opinion but I stuck to this because it was giving equivalent results.

Something like this maybe:

```python
def _has_common_variables(factor1, factor2):
    """Helper function to check if two polynomial factors share any variables"""
    return bool(set(factor1.gens).intersection(factor2.gens))
```

Reply:
That seems better.